tools/convert.py

```diff
@@ -97,47 +135,50 @@ def handle_tensors(args, writer, state_dict):
-        
-        elif n_params <= 1024:
-            # very small tensors
-            data_qtype = gguf.GGMLQuantizationType.F32
-        
-        elif n_dims == 4:
-            if min(data.shape[:2]) == 4: # output tensor
-                data_qtype = gguf.GGMLQuantizationType.F16
-            elif data_shape[-1] == 3: # 3x3 kernel
-                data_qtype = gguf.GGMLQuantizationType.F16
-            elif data_shape[-1] == 1: # 1x1 kernel
-                #data = np.squeeze(data) # don't do this
-                data_qtype = gguf.GGMLQuantizationType.F16
+        }
+
+        if old_dtype in (torch.float32, torch.bfloat16):
+            if n_dims == 1:
+                # one-dimensional tensors should be kept in F32
+                # also speeds up inference due to not dequantizing
+                data_qtype = gguf.GGMLQuantizationType.F32
+
+            elif n_params <= QUANTIZATION_THRESHOLD:
+                # very small tensors
+                data_qtype = gguf.GGMLQuantizationType.F32
+
+            elif ".weight" in key and any(x in key for x in blacklist):
+                data_qtype = gguf.GGMLQuantizationType.F32
+
+        if (    n_dims > 1                              # Skip one-dimensional tensors
+            and n_params >= REARRANGE_THRESHOLD         # Only rearrange tensors meeting the size requirement
+            and (n_params / 256).is_integer()           # Rearranging only makes sense if total elements is divisible by 256
+            and not (data.shape[-1] / 256).is_integer() # Only need to rearrange if the last dimension is not divisible by 256
+        ):
+            orig_shape = data.shape
+            data = data.reshape(n_params // 256, 256)
+            tensor_fixup[key] = orig_shape
 
         try:
             data = gguf.quants.quantize(data, data_qtype)
-        except gguf.QuantError as e:
-            tqdm.write(f"falling back to F16: {e}")
-            data_qtype = gguf.GGMLQuantizationType.F16
-            data = gguf.quants.quantize(data, data_qtype)
-        except AttributeError as e:
+        except (AttributeError, gguf.QuantError) as e:
             tqdm.write(f"falling back to F16: {e}")
             data_qtype = gguf.GGMLQuantizationType.F16
             data = gguf.quants.quantize(data, data_qtype)
 
         new_name = key # do we need to rename?
 
         shape_str = f"{{{', '.join(str(n) for n in reversed(data.shape))}}}"
-        tqdm.write(f"{f'%-{max_name_len}s' % f'{new_name}'} {old_dtype} --> {data_qtype.name}, shape = {shape_str}")
+        tqdm.write(f"{f'%-{max_name_len + 4}s' % f'{new_name}'} {old_dtype} --> {data_qtype.name}, shape = {shape_str}")
 
         writer.add_tensor(new_name, data, raw_dtype=data_qtype)
 
+    if not tensor_fixup:
+        return
+
+    fixup_str = "\n".join(" ".join((k,)+tuple(str(dimval) for dimval in v)) for k,v in tensor_fixup.items())
+    writer.add_string("comfy.gguf.tensor_fixup", fixup_str)
+

```

Comment:
You know, I was wondering. What if we add a separate fixup str for each key that needed to be fixed instead of a single string. We *may* need to adjust the max key length in llama.cpp for that but I think it may be nicer/easier to see when you look at the metadata on some place like HF.

Then you can just do something like this for each key:

```py
writer.add_string(f"comfy.gguf.orig_shape.{key}",  str(orig_shape))
```
```py
orig_shape = reader.get_field(f"comfy.gguf.orig_shape.{key}") # in a try/catch
```

Reply:
>  The key names aren't meant to store attributes/information like that IMO.

not to be argumentative, but isn't `f"comfy.gguf.orig_shape.{key}"` also storing metadata in the key name? except rather than it being the shape information directly, the metadata is the original _key_ instead of the original _shape_.

i'll admit i didn't give any consideration to stuff like how it will get displayed on HF though so that may be enough of a reason by itself.

> This is a valid concern, though that key name is under an `ifndef` so they can probably rebuild with `-DGGML_MAX_NAME=128` ..?

that's true, however the error you get is not at all intuitive. it's something like the tensor data isn't within the file (i forget the exact wording). it's definitely not anything along the lines of "the tensor name exceeds what was defined in `GGML_MAX_NAME`". took me a while to figure out what was happening, i originally though it just didn't like the characters i was using in the name.

***

anyway, seems like the verdict is to change it to the `comfy.gguf.orig_shape.ORIG_KEY` format? juuust double checking before i go ahead and mess with it.

Reply:
>not to be argumentative, but isn't f"comfy.gguf.orig_shape.{key}" also storing metadata in the key name

Touche.

There is also a `writer.add_array` function which would allow us to not do the whole string to list thing. I guess that's also another reason to do it this way though? It just takes/returns a list of ints as far as I can tell.

>how it will get displayed on HF

I'd assume the same way as the chat template on models, which looks pretty stupid without the newlines resolved sadly:

![image](https://github.com/user-attachments/assets/86ae98c7-bc72-45b7-905c-25b716e0c941)

>anyway, seems like the verdict is to change it to the comfy.gguf.orig_shape.ORIG_KEY format? juuust double checking before i go ahead and mess with it.

Yep I think that's for the best.

Reply:
how about this?

Reply:
Looks good. Any reason to use strings instead of `writer.add_array`?

Reply:
> Any reason to use strings instead of writer.add_array?

i haven't messed with arrays. it would complicate extracting the values/checking the types, you also can't control what the internal type is for array items: GGUF will take a guess. looks like for `int` it will use a signed 32bit type for items. not sure how likely it is, but that means if a tensor's original dimension is over 2,147,483,648 stuff will break.

do you need me to change it?

Reply:
Not sure if this would break anything else but to me it looks like converting it back is a lot cleaner/easier when it's an array:

```py
test = reader.get_field("tokenizer.ggml.token_type")
# check if exists
shape = torch.Size(test.data)
```

I think if we hit the 32 bit int limit in one dimension on a tensor we'll have more pressing things to worry about.

Reply:
i don't think your example would work, `.data` in a `ReaderField` doesn't lay the data out like that. also `data` isn't the the data itself, it's a list of indexes into `parts` for the parts of data that might make up a field. for example, an array will have an item for the key length, then the key data, then the array length, then the array data.

i will change it, but the code is going to be more complicated and possibly more fragile also.

Reply:
Huh, that example worked with the random LLM model I read the attributes out of and returned a valid object.

Reply:
Okay the tokenizer stuff must be added in some other way, let me look at it lol.

Reply:
Nvm I'm stupid it really was just the memory addresses lol.

Reply:
There must be a proper way to use it though if the web one can read it just fine.

![image](https://github.com/user-attachments/assets/4d2f4a77-96fa-4511-9215-01ab0ff7f35c)

Reply:
yeah, it's weird because a field with `320, 4, 3, 3` actually looks like this:

```plaintext
ReaderField(offset=144, 
  name='comfy.gguf.orig_shape.input_blocks.0.0.weight', 
  parts=[
    memmap([45], dtype=uint64), <-- length of key
    memmap([ 99, 111, 109, 102, 121,  46, 103, 103, 117, 102,  46, 111, 114, 105, 103,  95, 115, 104,  97, 112, 101,  46, 105, 110, 112, 117, 116,  95,  98, 108, 111,  99, 107, 115,  46,  48,  46,  48,  46, 119, 101, 105, 103, 104, 116], dtype=uint8), <-- key data
    memmap([9], dtype=uint32), <-- type (ARRAY)
    memmap([5], dtype=uint32), <-- subtype (INT32)
    memmap([4], dtype=uint64), <-- array length
    memmap([320], dtype=int32), <-- array item 0 (index 5 in parts)
    memmap([4], dtype=int32), 
    memmap([3], dtype=int32), 
    memmap([3], dtype=int32) <-- array item 3 (index 8 in parts)
  ], 
  data=[5, 6, 7, 8], <-- indexes into parts for the actual data
  types=[<GGUFValueType.ARRAY: 9>, <GGUFValueType.INT32: 5>] <-- types that comprise the field
)
```

anyway, i changed it. took a while to figure out since arrays are quite weird to deal with.

Reply:
You're the man. Now that's what I call clean. I'll give it a test, finish up the C++ parts and merge it if I don't find any issues.

Reply:
sounds good! let me know if you need any other changes/have questions/etc.

Reply:
Okay, I've pushed a completely rewritten convert patch which touches as little of the llama.cpp stuff as possible. As far as I can tell it's 1:1 to the old one with flux (sidenote, kind of amazing that the llama.cpp team managed to get this to be *literally* bit for bit reproducible lol.)

![image](https://github.com/user-attachments/assets/42be495e-616a-4b97-81d5-acafd5c563e3)

If you want you can pull in the changes here (maybe test them) and we can merge. If not I'll just resolve the merge conflict on the patch by keeping my one.

Reply:
tested with SD1.5, seems to work. small difference in sizes:

```plaintext
=== BEFORE
llama_model_quantize_internal: model size  =  1639.41 MB
llama_model_quantize_internal: quant size  =   467.09 MB

=== AFTER
llama_model_quantize_internal: model size  =  1639.41 MB
llama_model_quantize_internal: quant size  =   503.95 MB
```

there are some compile warnings but i think they're benign.

```plaintext
src/llama.cpp: In function ‘ggml_type img_tensor_get_type(quantize_state_internal&, ggml_type, const ggml_tensor*, llama_ftype)’:
src/llama.cpp:15871:33: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
15871 |         (arch == LLM_ARCH_FLUX) && (
      |         ~~~~~~~~~~~~~~~~~~~~~~~~^~~~
15872 |             (name.find("img_in.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15873 |             (name.find("time_in.in_layer.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15874 |             (name.find("vector_in.in_layer.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15875 |             (name.find("guidance_in.in_layer.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15876 |             (name.find("final_layer.linear.") != std::string::npos)
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15877 |         ) || (arch == LLM_ARCH_SD1 || arch == LLM_ARCH_SDXL) && (
      |         ~                        
src/llama.cpp:15885:33: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
15885 |         (arch == LLM_ARCH_FLUX) && (
      |         ~~~~~~~~~~~~~~~~~~~~~~~~^~~~
15886 |             (name.find("txt_in.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15887 |             (name.find("time_in.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15888 |             (name.find("vector_in.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15889 |             (name.find("guidance_in.") != std::string::npos) ||
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15890 |             (name.find("final_layer.") != std::string::npos)
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15891 |         ) || (arch == LLM_ARCH_SD1 || arch == LLM_ARCH_SDXL) && (
      |         ~                        
src/llama.cpp: In function ‘llama_rope_type llama_rope_type(const llama_model*)’:
src/llama.cpp:17589:12: warning: enumeration value ‘LLM_ARCH_FLUX’ not handled in switch [-Wswitch]
17589 |     switch (model->arch) {
      |            ^
src/llama.cpp:17589:12: warning: enumeration value ‘LLM_ARCH_SD1’ not handled in switch [-Wswitch]
src/llama.cpp:17589:12: warning: enumeration value ‘LLM_ARCH_SDXL’ not handled in switch [-Wswitch]
```

looking at the heuristic part, not really sure what it's complaining about. the rope type thing is probably unreachable code for this kind of model.

i synchronized with your changes.

Reply:
Size slightly increased but I think it's worth it. For me it made it a lot more coherent.

Odd how mine doesn't complain about either. I've removed the rope stuff since we never hit that switch as far as I can tell but didn't get a warning there either.

![image](https://github.com/user-attachments/assets/37481890-5658-49c3-817d-b4e6aa509ddd)

Reply:
probably just a different C++ compiler version (i'm on GCC 14.2.1 which is probably more recent than most people). also compiling with `make` rather than `cmake` could be a factor.

not a problem or anything, just thought i'd mention it.

> Size slightly increased but I think it's worth it. For me it made it a lot more coherent.

weirdly i got worse results. :) just the luck of the seed change lottery. or maybe i need to keep my SD models from getting too smart by smashing down the weights.

***

unrelated question that probably shouldn't be here: is there any conceivable way to generate an importance matrix for these models? i have no idea how that would work.

Reply:
>weirdly i got worse results. :) just the luck of the seed change lottery. 

We can always adjust the weight logic, nothing is set it stone if it turns out to be worse in some way. It's mostly just the embedding keys that I added since without those it was ignoring most of the prompt or falling apart at Q3_K_M lol.

>is there any conceivable way to generate an importance matrix for these models?
>i have no idea how that would work.

That makes two of us :,)
I still have to look into that, probably with T5 first since that's supported by llama.cpp but doesn't work with imatrix either.